[WIP] notify user when mentioned in comments

blizzz opened the pull request:
I underestimated the effort  of understand and working with Notifications :blush: however this now is a working prototype. Without any GUI helpers though, i.e. atm you'd need to know the username.

contributes to #24445 

Some TODOs left:
- [ ] adjust unit tests
- [ ] write unit tests
- [x] don't notifiy yourself if you mentioned yourself
- [ ] get rid off @NoCSRFRequired in the controller
- [ ] delete notifications when a comment is deleted

@nickvergessen if you like to have a glance on Notifications usage?

blizzz committed:
(WIP) notify user when mentioned in comments

nickvergessen commented on lib/private/Comments/Manager.php:
```diff
@@ -750,4 +777,31 @@ public function deleteReadMarksOnObject($objectType, $objectId) {
+				->setObject('comment', $comment->getId())
+				->setSubject('mention', [ $comment->getObjectType(), $comment->getObjectId() ])
+				->setDateTime($comment->getCreationDateTime())
+				->setLink($this->urlGenerator->linkToRoute(
```
linkToRouteAbsolute

nickvergessen commented on apps/comments/appinfo/application.php:
```diff
@@ -0,0 +1,49 @@
+<?php
+
+/**
+ * ownCloud - Activity App
```
:speak_no_evil:

blizzz replied:
:blush:

nickvergessen commented on apps/comments/appinfo/application.php:
```diff
@@ -0,0 +1,49 @@
+use OCP\AppFramework\App;
+use OCP\IContainer;
+
+class Application extends App {
```
File should be in lib/AppInfo/
with PSR-4 naming

nickvergessen commented on apps/comments/controller/notifications.php:
```diff
@@ -0,0 +1,122 @@
+ *
+ */
+
+namespace OCA\Comments\Controller;
```
PSR-4 lib/Controller/Notifications.php

nickvergessen commented on lib/private/Comments/Manager.php:
```diff
@@ -750,4 +777,31 @@ public function deleteReadMarksOnObject($objectType, $objectId) {
 		return ($affectedRows > 0);
 	}
+
+	protected function triggerNotifications(IComment $comment) {
```
Now the manager creates the notifications, but the app is required to display them.
I guess you could also listen to an event in the app and then add the notifications from there.

blizzz replied:
You're right, this would make it a cleaner approach.

blizzz committed:
Fix doc, and create absolute URL for as notification link.

blizzz committed:
PSR-4 compatibility changes

blizzz commented:
@nickvergessen addressed all your points :)

blizzz committed:
also move notification creation to comments app

blizzz committed:
Do not notify yourself

nickvergessen commented on apps/comments/appinfo/app.php:
```diff
@@ -49,5 +49,27 @@ function() {
+		);
+	},
+	function () {
+		return ['id' => 'comments', 'name' => 'Comments'];
```
name should be translated

nickvergessen commented on apps/comments/lib/Controller/Notifications.php:
```diff
@@ -0,0 +1,122 @@
+			if($comment->getObjectType() !== 'files') {
+				return new Response();
+			}
+			$files = $this->folder->getById($comment->getObjectId());
```
Are we sure `Filesystem::init()` has been called before? I had to do that in my background job manually, but I guess since you are on a loggin in page, it is?

blizzz replied:
Yes… well, I did not re-check the code path, but all authentication and filesystem stuff should have happened. I also tried to access the controller page being logged out. Well, you just need to log in and again, and then everything works just as well.

nickvergessen commented on apps/comments/lib/Notification/Listener.php:
```diff
@@ -0,0 +1,96 @@
+			$notification = $this->notificationManager->createNotification();
+			$notification
+				->setApp('comments')
+				->setUser($user)
```
everything apart from `setUser()` and `$this->notificationManager->notify()` could be done outside the loop, so you only do it once, but I guess it doesn't really matter because no one is goign to mention hundreds of people with the 1k character limit

blizzz replied:
good point. also no need to waste cycles unnecessarily, even if they are cheap.

nickvergessen commented on apps/comments/lib/Notification/Notifier.php:
```diff
@@ -0,0 +1,102 @@
+			if(!is_null($commenter)) {
+				$displayName = $commenter->getDisplayName();
+			} else {
+				$displayName = $l->t('an unknown user');
```
I'd use the ID instead?

nickvergessen replied:
Also because string concatenation is bad with translations and you inject it into `You were mentioned in a comment on "%s" by %s`

blizzz replied:
I would not use the ID, because the user can be deleted and recreated. However, as i remember now again, we deal with this on comment side and have special actor type "deleted_users". So, using the ID would be ok as fallback in this case, but the string concatenation problem would pop up anyway in a different part. Other translation system have an optional context parameter that helps in cases like this.

nickvergessen commented:
Looks good and works :tada:

blizzz committed:
unit test for controller and application

nickvergessen commented on apps/comments/appinfo/app.php:
```diff
@@ -20,6 +20,8 @@
  *
  */
 
+$l = \OC::$server->getL10N('comments');
```
You can do that in the closure, so we save it as long as possible?

blizzz replied:
:ok_hand:

blizzz committed:
smaller fixes

- translatable app name
- remove doubles in mention array
- micro perf optimization
- display name: special label for deleted users, keep user id for users that could not be fetched from userManager

blizzz committed:
Comment Notification-Listener Unit Test

PVince81 commented:
We're past feature freeze.

@MTRichards @cmonteroluque maybe something to schedule at some point.

Will move to 9.2 for now.

PVince81 commented:
Is there still interest in getting this feature finished ? Should someone take over ?

PVince81 commented:
Backlog for now, but possible future feature for comments @pmaier1

pmaier1 commented:
I would really like to have more possibilities for notifications from oC. This would make UX more vibrant and enhance collaboration. Is there/should we start a list for possible notifications?
